nuttx-patches/00003-PENDING-add-set-ex-to-nsh.patch

```diff
@@ -0,0 +1,314 @@
+       pstate->cn_outstream       = OUTSTREAM(pstate);
+ #endif
+     }
++#ifndef CONFIG_NSH_DISABLESCRIPT
++  /* Set the initial option flags */
+ 
++  pstate->cn_vtbl.np.np_flags = NSH_NP_SET_OPTIONS_INIT;
++#endif
+   return pstate;
+ }
+diff --git NuttX/apps/nshlib/nsh_consolemain.c NuttX/apps/nshlib/nsh_consolemain.c
+index f89944f..2797610 100644
+--- NuttX/apps/nshlib/nsh_consolemain.c
++++ NuttX/apps/nshlib/nsh_consolemain.c
+@@ -85,6 +85,11 @@ int nsh_consolemain(int argc, char *argv[])
+   /* Execute the start-up script */
+ 
+   (void)nsh_initscript(&pstate->cn_vtbl);
++#ifndef CONFIG_NSH_DISABLESCRIPT
++  /* Reset the option flags */
++
++  pstate->cn_vtbl.np.np_flags = NSH_NP_SET_OPTIONS_INIT;
++#endif
+ #endif
+ 
+ #ifdef CONFIG_NSH_USBDEV_TRACE
+diff --git NuttX/apps/nshlib/nsh_envcmds.c NuttX/apps/nshlib/nsh_envcmds.c
+index cbda78b..277665b 100644
+--- NuttX/apps/nshlib/nsh_envcmds.c
++++ NuttX/apps/nshlib/nsh_envcmds.c
+@@ -311,29 +311,84 @@ int cmd_pwd(FAR struct nsh_vtbl_s *vtbl, int argc, char **argv)
+  * Name: cmd_set
+  ****************************************************************************/
+ 
+-#ifndef CONFIG_DISABLE_ENVIRON
+ #ifndef CONFIG_NSH_DISABLE_SET
+ int cmd_set(FAR struct nsh_vtbl_s *vtbl, int argc, char **argv)
+ {
+   FAR char *value;
+-  int ret;
++  int ret = OK;
++  int ndx = 1;
++#ifndef CONFIG_NSH_DISABLESCRIPT
++  FAR char *popt;
++  const char opts[] = NSH_NP_SET_OPTIONS;
++  int op;
+ 
+-  /* Trim whitespace from the value */
++  /* Support set [{+|-}{e|x|xe|ex}] [<name> <value>] */
+ 
+-  value = nsh_trimspaces(argv[2]);
++  if (argc == 2 || argc == 4)
++    {
++      if (strlen(argv[1]) < 2)
++        {
++          ret = -EINVAL;
++          nsh_output(vtbl, g_fmtargrequired, argv[0], "set", NSH_ERRNO);
++        }
++      else
++        {
++          op = argv[1][0];
++          if (op != '-' && op != '+')
++            {
++              ret = -EINVAL;
++              nsh_output(vtbl, g_fmtarginvalid, argv[0], "set", NSH_ERRNO);
++            }
++          else
++            {
++              value = &argv[1][1];
++              while(*value && *value != ' ')
++                {
++                  popt = strchr(opts, *value++);
++                  if (popt == NULL)
++                    {
++                      nsh_output(vtbl, g_fmtarginvalid, argv[0], "set", NSH_ERRNO);
++                      ret = -EINVAL;
++                      break;
++                    }
++                  if (op == '+')
++                    {
++                      vtbl->np.np_flags |= 1 << (popt-opts);
++                    }
++                  else
++                    {
++                      vtbl->np.np_flags &= ~(1 << (popt-opts));
++                    }
++                }
++              if (ret == OK)
++                {
++                  ndx = 2;
++                }
++            }
++         }
++      }
++#  ifndef CONFIG_DISABLE_ENVIRON
++  if (ret == OK && (argc == 3 || argc == 4))
```

Comment:
I got really confused by the ifdefs and indentation here but I came to the conclusion that it must be correct.